### PR TITLE
Create lock to avoid race conditions when adding tokens

### PR DESCRIPTION
**Pull request by yabirgb**
It happened that during the migration process two tasks tried to add the same token to the database. The sequence was:

1. greenlet A checks existence of token and fails
2. greenlet B checks existence of token and fails
3. greenlet A adds the token
4. greenlet B tries to add the token but it already exists

The sections that checks existence and adds the token is critic and should be protected

### CODE REVIEW

File: rotkehlchen/assets/utils.py [hunk 1]
```diff
@@ -68,36 +68,43 @@ def get_or_create_evm_token(
...
-            protocol=protocol,
-            underlying_tokens=underlying_tokens,
-        )
-        # This can but should not raise InputError since it should not already exist
-        ethereum_token = add_ethereum_token_to_db(token_data)
-        with userdb.user_write() as cursor:
-            userdb.add_asset_identifiers(cursor, [ethereum_token.identifier])
+    # Protect the whole section with a lock to avoid the case where the try is executed
+    # by two greenlets, both enter the except and then they fail. The existence check and addition
+    # is a critic section.
```
Comment (LefterisJP) [inline rotkehlchen/assets/utils.py:73]:
```suggestion
    # is a critical section.
```

File: rotkehlchen/assets/utils.py [hunk 2]
```diff
@@ -68,36 +68,43 @@ def get_or_create_evm_token(
...
+                chain=chain,
+                token_kind=token_kind,
+                name=name,
+                decimals=decimals,
+                symbol=symbol,
+                protocol=protocol,
+                underlying_tokens=underlying_tokens,
+            )
+            # This can but should not raise InputError since it should not already exist
+            ethereum_token = add_ethereum_token_to_db(token_data)
```
Comment (LefterisJP) [inline rotkehlchen/assets/utils.py:105]: My comment here is that as far as I can see the only place in this function when context switch can happen is these 2 DB functions. (I don't think that logging functions can context switch, right?)

So instead of adding a new lock you can perhaps re-introduce (seems it's removed?) the critical section in the DB cursor creation that puts `set_progress_handler` to `0` and resets it to the last value once out of the section?

This approach may remove the need for the extra lock.

So I mean to reverts this commit: 339244579bec8c2c97d37ac01a2c8190f89bd3a7 except for the actual usage. So except for using `self.enter_critical_section()` and `self.exit_critical_section()`.

Instead you can use those here in this function by making a different ctx() method for critical section and using that ctx() in this function that enters and exits critical section.

Does that make sense?

You would need to do that for both the global db write at `add_ethereum_token_to_db` and userdb write with `add_asset_identifiers`.

Comment (yabirgb) [reply rotkehlchen/assets/utils.py:105]: I've doing a bit of research and seems that it possible to make the logs async using https://github.com/gevent/gevent/blob/beb98ce4742181caaa6cec9cedecaaa2a04f910b/src/gevent/os.py#L28 but is not the default afaik and we don't do that.

Comment (yabirgb) [reply rotkehlchen/assets/utils.py:105]: but I believe your proposal doesn't solve the issue because the problem is not writing at the same time in the database. The problem is that the two operations of checking that a token exists or not and the process of adding it to the database if not shouldn't be executed by two greenlets at the same time. Both the read and the write should be "atomic"

Comment (LefterisJP) [reply rotkehlchen/assets/utils.py:105]: I am not saying to make the logs async.

I am saying that so long as there is nothing that will context switch the entire function out of the current greenlet then the function will indeed in fact be atomic.

Comment (yabirgb) [reply rotkehlchen/assets/utils.py]: Okey got you. My mind was thinking in threads and not in greenlets. I believe is correct now

File: rotkehlchen/assets/utils.py [hunk 3]
```diff
@@ -20,13 +20,12 @@
...
 
 
-def add_ethereum_token_to_db(token_data: EvmToken) -> EvmToken:
+def add_ethereum_token_to_db(token_data: EvmToken, globaldb: 'GlobalDBHandler') -> EvmToken:
     """Adds an ethereum token to the DB and returns it
 
     May raise:
     - InputError if token already exists in the DB
     """
-    globaldb = GlobalDBHandler()
```
Comment (LefterisJP) [inline rotkehlchen/assets/utils.py:29]: why this change? Why in one function create it and in the other pass it? I think this is not a good change. I would revert this and keep the function's interface simple.

**Review by LefterisJP — APPROVED** [review]
Looks good to me despite the one comment on the interface. I would like you or someone else to change it in another PR since this one is green.
